pkg/kubelet/kubelet.go

```diff
@@ -3282,7 +3311,21 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
 
 func (kl *Kubelet) generateAPIPodStatus(pod *api.Pod, podStatus *kubecontainer.PodStatus) api.PodStatus {
 	glog.V(3).Infof("Generating status for %q", format.Pod(pod))
+
+	// any pod that requires eviction must be added to the list to sync.
+	attrs := &lifecycle.PodEvictAttributes{Pod: pod}
+	for _, podEvictor := range kl.podEvictors {
+		if result := podEvictor.Evict(attrs); result.Evict {
```

Comment:
@Random-Liu - In the current code, the pod is set to failed when we invoke `generateAPIPodStatus` in `syncPod` and we set that in the status manager.  I think I like the idea of `syncPod` calling `killPod` if it sees the `apiStatus` has transitioned to a failure state.  This means that we won't require the housekeeper to kill the pod.

From an interface definition perspective, we could have `PodEvictor` have a `SyncPod` method that returns true if the pod should be synch-ed, and keep that separate from the `Evict` method that is then only invoked from `generateAPIPodStatus` during normal `syncPod` invocation.

/cc @vishh WDYT?  In that way, we can know that `Evict` will not be called multiple times for the same pod.